File: src/core/node/for_node.ts
import { Position } from "../../error/position.ts";
import { Nodes, TYPES } from "../../utils/constants.ts";
import { Token } from "../token.ts";

export class ForNode {
  public type: TYPES;
  public positionStart: Position;
  public positionEnd: Position;

  constructor(
    public varNameToken: Token,
    public startValue: Nodes,
    public endValue: Nodes,
    public bodyNode: Nodes,
    public stepValueNode?: Nodes,
  ) {
    this.type = bodyNode.type;
    this.positionStart = varNameToken.positionStart!;
    this.positionEnd = bodyNode.positionEnd;
  }

  public represent(): string {
    return `FOR ${this.varNameToken.represent()} ${this.startValue.represent()} ${this.endValue.represent()} ${
      this.stepValueNode?.represent()
    } ${this.bodyNode.represent()}`;
  }
}

File: src/core/nodes/for_node.rs
use crate::core::token::Token;
use crate::utils::constants::Nodes;
use crate::utils::position::Position;

#[derive(Debug, Clone)]
pub struct ForNode {
    pub var_name_token: Token,
    pub start_value: Nodes,
    pub end_value: Nodes,
    pub body_node: Nodes,
    pub step_value_node: Option<Nodes>,
    pub pos_start: Position,
    pub pos_end: Position,
}

impl ForNode {
    pub fn new(
        var_name_token: Token,
        start_value: Nodes,
        end_value: Nodes,
        body_node: Nodes,
        step_value_node: Option<Nodes>,
    ) -> ForNode {
        let pos_start = var_name_token.clone().pos_start;
        let pos_end: Position;
        match body_node.clone() {
            Nodes::Number(node) => pos_end = node.pos_end,
            Nodes::StringNode(node) => pos_end = node.pos_end,
            Nodes::CharNode(node) => pos_end = node.pos_end,
            Nodes::BinOp(node) => pos_end = node.pos_end,
            Nodes::IfNode(node) => pos_end = node.pos_end,
            Nodes::UnaryOp(node) => pos_end = node.pos_end,
            Nodes::ForNode(node) => pos_end = node.pos_end,
            Nodes::VarAssignNode(node) => pos_end = node.pos_end,
            Nodes::VarReassignNode(node) => pos_end = node.pos_end,
            Nodes::VarAccessNode(node) => pos_end = node.pos_end,
            Nodes::WhileNode(node) => pos_end = node.pos_end,
            Nodes::BooleanNode(node) => pos_end = node.pos_end,
            Nodes::FunDef(node) => pos_end = node.pos_end,
            Nodes::CallNode(node) => pos_end = node.pos_end,
        };

        ForNode {
            var_name_token,
            start_value,
            end_value,
            body_node,
            step_value_node,
            pos_start,
            pos_end,
        }
    }
}
